src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs

```diff
@@ -127,30 +127,59 @@ private static void ProcessOperator(Stack<FilterExpression> filterStack, Operato
             properties = Enumerable.Empty<string>();
         }
 
-        bool valid = false;
-        if (_condition != null)
+        FilterExpression? current = this;
+        Stack<FilterExpression> filterStack = new();
+        Stack<string[]?> result = new();
+        do
         {
-            valid = _condition.ValidForProperties(properties, propertyProvider);
-            if (!valid)
+            //Push root's right child and then root to stack then Set root as root's left child
+            while (current != null)
             {
-                invalidProperties = new string[1] { _condition.Name };
+                if (current._right != null)
+                {
+                    filterStack.Push(current._right);
+                }
+                filterStack.Push(current);
+                current = current._left;
             }
-        }
-        else
-        {
-            invalidProperties = _left!.ValidForProperties(properties, propertyProvider);
-            var invalidRight = _right!.ValidForProperties(properties, propertyProvider);
-            if (null == invalidProperties)
+            //If the popped item has a right child and the right child is at top of stack, then remove the right child from stack, push the root back and set root as root's right child.
+            current = filterStack.Pop();
+            if (filterStack.Count > 0 && current._right == filterStack.First())
```

Comment:
Could we use [`Peek`](https://docs.microsoft.com/dotnet/api/system.collections.stack.peek) instead of `First`? It wasn't clear if `First` was starting from top or bottom of the `Stack`.